nox/manifest.py

```diff
@@ -0,0 +1,166 @@
+    run by nox.
+
+    It is possible for this to be mutated during execution. This allows for
+    useful use cases, such as for one session to "notify" another.
+
+    Args:
+        session_functions (Mapping[str, function]): The registry of discovered
+            session functions.
+        global_config (): The global configuration.
+    """
+    def __init__(self, session_functions, global_config):
+        self._all_sessions = []
+        self._config = global_config
+
+        # Create the sessions based on the provided session functions.
+        for name, func in session_functions.items():
+            self.make_session(name, func)
+
+        # By default, every available session starts off in the queue.
+        self._queue = copy.copy(self._all_sessions)
+        self._consumed = []
+
+        # If the configuration specified filters, apply those now.
+        if self._config.sessions:
+            self.filter_by_name(self._config.sessions)
+        if self._config.keywords:
+            self.filter_by_keywords(self._config.keywords)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """Return the next item in the queue.
+
+        Raises:
+            StopIteration: If the queue has been entirely consumed.
+        """
+        if not len(self._queue):
+            raise StopIteration
+        session = self._queue.pop(0)
+        self._consumed.append(session)
+        return session
+
+    def __nonzero__(self):
+        return bool(self._queue) or bool(self._consumed)
+
+    def filter_by_name(self, specified_sessions):
+        """Filter sessions in the queue based on the user-specified names.
+
+        Args:
+            specified_sessions (Sequence[str]): A list of specified
+                session names.
+        """
+        # Filter the sessions remaining in the queue based on
+        # whether they are individually specified.
+        self._queue = [x for x in self._queue if (
+            x.name in specified_sessions or
+            x.signature in specified_sessions)]
+
+        # If a session was requested and was not found, complain loudly.
+        missing_sessions = set(specified_sessions) - set(
+            itertools.chain(
+                [x.name for x in self._all_sessions if x.name],
+                [x.signature for x in self._all_sessions if x.signature]))
+        if missing_sessions:
+            logger.error('Sessions {} not found.'.format(', '.join(
+                missing_sessions)))
+            return False
+
+    def filter_by_keywords(self, keywords):
+        """Filter sessions using pytest-like keyword expressions.
+
+        Args:
+            keywords (Collection[str]): A collection of keywords which
+                session names are checked against.
+        """
+        self._queue = [
+            x for x in self._queue
+            if keyword_match(keywords, [x.signature or x.name])]
+
+    def make_session(self, name, func):
```

Comment:
I can change this. Would it be okay if we merged this and I will follow up with that PR? Just so it does not get even longer?